elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/queryengines/sql/SQLQueryConstructor.java

```diff
@@ -0,0 +1,507 @@
+import com.yahoo.elide.core.exceptions.InvalidPredicateException;
+import com.yahoo.elide.core.filter.FilterPredicate;
+import com.yahoo.elide.core.filter.FilterTranslator;
+import com.yahoo.elide.core.filter.expression.FilterExpression;
+import com.yahoo.elide.core.filter.expression.PredicateExtractionVisitor;
+import com.yahoo.elide.core.sort.Sorting;
+import com.yahoo.elide.datastores.aggregation.AggregationDictionary;
+import com.yahoo.elide.datastores.aggregation.metadata.metric.MetricFunctionInvocation;
+import com.yahoo.elide.datastores.aggregation.metadata.models.Table;
+import com.yahoo.elide.datastores.aggregation.metadata.models.TimeDimension;
+import com.yahoo.elide.datastores.aggregation.metadata.models.TimeDimensionGrain;
+import com.yahoo.elide.datastores.aggregation.query.ColumnProjection;
+import com.yahoo.elide.datastores.aggregation.query.Query;
+import com.yahoo.elide.datastores.aggregation.query.TimeDimensionProjection;
+import com.yahoo.elide.datastores.aggregation.queryengines.sql.annotation.FromSubquery;
+import com.yahoo.elide.datastores.aggregation.queryengines.sql.annotation.FromTable;
+import com.yahoo.elide.datastores.aggregation.queryengines.sql.annotation.JoinTo;
+import com.yahoo.elide.datastores.aggregation.queryengines.sql.metadata.SQLAnalyticView;
+import com.yahoo.elide.datastores.aggregation.queryengines.sql.metadata.SQLColumn;
+import com.yahoo.elide.datastores.aggregation.queryengines.sql.query.SQLQueryTemplate;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * Class to construct query template into real sql query
+ */
+public class SQLQueryConstructor {
+    private final AggregationDictionary dictionary;
+
+    public SQLQueryConstructor(AggregationDictionary dictionary) {
+        this.dictionary = dictionary;
+    }
+
+    /**
+     * Construct sql query with a template and sorting, where and having clause.
+     *
+     * @param clientQuery original query object
+     * @param template query template constructed from client query
+     * @param sorting sorting clause
+     * @param whereClause where clause
+     * @param havingClause having clause
+     * @return constructed SQLQuery object contains all information above
+     */
+    public SQLQuery resolveTemplate(Query clientQuery,
+                                    SQLQueryTemplate template,
+                                    Sorting sorting,
+                                    FilterExpression whereClause,
+                                    FilterExpression havingClause) {
+        SQLAnalyticView sqlAnalyticView = (SQLAnalyticView) clientQuery.getTable();
+        Table elideTable = clientQuery.getTable();
+        Class<?> tableCls = clientQuery.getTable().getCls();
+        String tableAlias = getClassAlias(tableCls);
```

Comment:
getClassAlias can move to the Table instead of a static function associated with the dictionary (as part of the refactoring to not sublcass the dictionary).

Reply:
👍